File: Level 7 - System Design & Probabilistic DS/cuckoo_filter/advanced_problems.cpp
#include <iostream>
#include <string>
#include <vector>
#include <iomanip>

/**
 * Advanced Cuckoo Filter Problems in C++
 * Focuses on Load Factor and Saturation.
 */

#include "basic_operations.cpp"

int main() {
    int capacity = 1000;
    CuckooFilter cf(capacity);

    std::cout << "Filling filter until it fails..." << std::endl;
    int count = 0;
    for (int i = 0; i < 2000; ++i) {
        if (!cf.insert("val_" + std::to_string(i))) {
            count = i;
            break;
        }
    }

    double lf = (double)count / capacity * 100;
    std::cout << "Capacity: " << capacity << std::endl;
    std::cout << "Inserted: " << count << std::endl;
    std::cout << "Load Factor: " << std::fixed << std::setprecision(2) << lf << "%" << std::endl;

    std::cout << "\nFreeing space by deleting 50 items..." << std::endl;
    for(int i=0; i<50; ++i) cf.remove("val_" + std::to_string(i));
    
    std::cout << "Insert after delete: " << (cf.insert("retry_1") ? "Success" : "Failed") << std::endl;

    return 0;
}

File: Level 7 - System Design & Probabilistic DS/cuckoo_filter/advanced_problems.java
import java.util.*;

/**
 * Advanced Cuckoo Filter Problems in Java
 * Focuses on Capacity Saturation and Load Factor analysis.
 */
public class advanced_problems {

    public static void main(String[] args) {
        int capacity = 1000;
        basic_operations filter = new basic_operations(capacity);

        System.out.println("Filling filter until insertion failure...");
        int count = 0;
        while (count < 2000) {
            if (!filter.insert("element_" + count)) {
                break;
            }
            count++;
        }

        double loadFactor = (double) count / capacity * 100;
        System.out.println("Insertion failed at count: " + count);
        System.out.println("Achieved Load Factor: " + String.format("%.2f", loadFactor) + "%");
        System.out.println("Note: Cuckoo filters typically achieve 90-95% load factor with 4-slot buckets.");

        System.out.println("\nTesting deletion to free up space...");
        for (int i = 0; i < 100; i++) {
            filter.delete("element_" + i);
        }

        System.out.println("Trying to insert after deletion...");
        System.out.println("Insert successful: " + filter.insert("new_element_999"));
    }
}
